File: Leetcode/C++/2024. Maximize the Confusion of an Exam/solution.cpp
// https://leetcode.com/problems/maximize-the-confusion-of-an-exam/description/

#include <vector>
#include <string>

using namespace std;

class Solution
{
public:
    int maxConsecutiveAnswers(string answerKey, int k)
    {
        return max(findMax(answerKey, k, 'T'), findMax(answerKey, k, 'F'));
    }
    // If diffKey == 'T' change 'T' to 'F'
    int findMax(string &answerKey, int k, char diffKey)
    {
        int left = 0, maxCount = 0, curK = k;
        for (int right = 0; right < answerKey.size(); right++)
        {
            if (answerKey[right] == diffKey)
            {
                curK--;
            }
            // Use all change, move left index to next (diffKey index + 1)
            while (curK < 0)
            {
                if (answerKey[left] == diffKey)
                {
                    curK++;
                }
                left++;
            }
            maxCount = max(maxCount, right - left + 1);
        }
        return maxCount;
    }
};

File: Leetcode/Python3/2024. Maximize the Confusion of an Exam/solution.py
# https://leetcode.com/problems/maximize-the-confusion-of-an-exam/description/

class Solution:  # Sliding window
    def maxConsecutiveAnswers(self, answerKey: str, k: int) -> int:
        return max(self.findMax(answerKey, k, "F"), self.findMax(answerKey, k, "T"))

    def findMax(self, answerKey: str, k: int, changeKey: str):
        maxCount = 0
        left, curK = 0, k
        for right, val in enumerate(answerKey):
            if val == changeKey:
                curK -= 1
            # If use all change chance move left index to next changeKey
            while curK < 0:
                if answerKey[left] == changeKey:
                    curK += 1
                left += 1
            maxCount = max(maxCount, right - left + 1)
        return maxCount
